ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/assertions/CommandDeduplicationAssertions.scala

```diff
@@ -0,0 +1,118 @@
+import com.google.protobuf.duration.{Duration => DurationProto}
+import io.grpc.Status.Code
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.math.Ordering.Implicits._
+
+object CommandDeduplicationAssertions {
+
+  def assertDeduplicationDuration(
+      requestedDeduplicationDuration: DurationProto,
+      previousSubmissionSendTime: Instant,
+      completionReceiveTime: Instant,
+      completion: Completion,
+      submittingParty: Party,
+      ledger: ParticipantTestContext,
+  )(implicit executionContext: ExecutionContext): Future[Unit] = {
+    val requestedDuration = DurationConversion.fromProto(requestedDeduplicationDuration)
+    ledger.features.commandDeduplicationFeatures.getDeduplicationPeriodSupport.durationSupport match {
+      case DurationSupport.DURATION_NATIVE_SUPPORT =>
+        val reportedDurationProto = assertDefined(
+          completion.deduplicationPeriod.deduplicationDuration,
+          "No deduplication duration has been reported",
+        )
+        val reportedDuration = DurationConversion.fromProto(reportedDurationProto)
+        assert(
+          reportedDuration >= requestedDuration,
+          s"The reported deduplication duration $reportedDuration was smaller than the requested deduplication duration $requestedDuration.",
+        )
+        Future.unit
+      case DurationSupport.DURATION_CONVERT_TO_OFFSET =>
+        val reportedOffset = assertDefined(
+          completion.deduplicationPeriod.deduplicationOffset,
+          "No deduplication offset has been reported",
+        )
+        if (completion.getStatus.code == Code.ALREADY_EXISTS.value) {
+          val completionStreamRequest = ledger.completionStreamRequest(
+            LedgerOffset.of(LedgerOffset.Value.Absolute(reportedOffset))
+          )(submittingParty)
+          ledger
+            .findCompletion(completionStreamRequest)(_.commandId == completion.commandId)
+            .map { optOffsetCompletionResponse =>
+              val _ = assertDefined(
+                optOffsetCompletionResponse,
+                s"No completion with the command ID '${completion.commandId}' after the reported offset $reportedOffset has been found",
+              )
+            }
+        } else {
+          // Let t2 (completionReceiveTime) be the time when the application received the completion offset that specifies the deduplication offset off1.
+          // Let t1 (previousSubmissionSendTime) be the time when the application sent off the submission that completed at off1.
+          // Then t2 - t1 >= requested deduplication duration.
+          assert(
+            requestedDuration <= Duration
+              .between(previousSubmissionSendTime, completionReceiveTime),
+            s"The requested deduplication duration $requestedDeduplicationDuration was greater than the duration between sending the previous submission and receiving the next completion.",
+          )
+          Future.unit
```

Comment:
This looks like dead code right now because all use cases of this method should have the status code `ALREADY_EXISTS`, so we never end up in this branch.

This assertion still does not make sense, because `previousSubmissionSendTime` refers to some arbitrary offset that was submitted previously by the test, which is therefore independent of `reportedOffset`. What would be the correct thing to do here is to remember a "submission time" for all offsets that were previously generated, and then extract the time for `reportedOffset` from this history and use that as `t1`. 

Since all this is irrelevant for this PR, I recommend to just remove this check and put a placeholder like `fail("This case is not yet supported by this assertion")` here so that nobody can accidentally run into this problem.

Reply:
Unfortunately, the entire duration -> offset conversion case is dead code, as this test is ignored for `bazel test //ledger/ledger-api-test-tool-on-canton:conformance-test`, as, among the others, Canton does not expose feature descriptors. So, to test it (locally), I had to remove the exclusion and comment out the code for other deduplication period support cases.

But I agree, I removed this check for now.